Release notes v0.3.0 (#265)
grafana/xk6-browser: release notes/v0.3.0.md

xk6-browser v0.3.0 is here! :tada:

This is a minor release with some important bug fixes, a couple of new features, improvements and code refactors!

## New features

- Implemented `Browser.on(event)`. ([#268](https://github.com/grafana/xk6-browser/pull/268), [#287](https://github.com/grafana/xk6-browser/pull/287))

  `on('disconnected')` returns a `Promise` you can wait on to get notified when the WebSocket connection to the browser is closed. See [this example](https://github.com/grafana/xk6-browser/blob/v0.3.0/examples/browser_on.js) for details.

  This is the first async method in xk6-browser, using the newly released [event loop support in k6](https://github.com/grafana/k6/pull/2228), and marks the start of our transition to async APIs. :tada:

- Exposed preconfigured mobile device emulation settings. ([#289](https://github.com/grafana/xk6-browser/pull/289))

  We maintain an internal list of popular devices with their viewport size, user agent and scaling factor, which is useful to test how a web site looks and behaves on different devices. See [this example](https://github.com/grafana/xk6-browser/blob/v0.3.0/examples/device_emulation.js) for details.

> 
> 

Author: ```suggestion
- Added the `executablePath` option. ([#246](https://github.com/grafana/xk6-browser/pull/246))

  You can now run a Chrome browser of your choice by providing the file path.
```

Reviewer: This is more of a fix, no? The option was already there, it just didn't work before.

Author: Many options and features are there, but they don't work :( Both are fine with me (feature or bug).

Reviewer: It's a feature that didn't work before, but now does, so a fix is more appropriate. Please mention it there.

Author: @imiric, 👍, does it make more sense now?

```suggestion
- Added the `executablePath` option. ([#246](https://github.com/grafana/xk6-browser/pull/246))

  Although you could fill this option out, it wasn't working before, and we were using a predefined path by default. We implemented the option code, and you can now run a Chrome browser of your choice by providing the file path.
```

Reviewer: Yes, as a fix.

Author: See: bbfe45

Author: See: https://github.com/grafana/xk6-browser/pull/265/commits/bbfe4532fc39466dc7103006a93a07b43fc49ebd

## Bugs fixed

- Made concealed elements clickable. ([#264](https://github.com/grafana/xk6-browser/pull/264))

>   Previously, interacting with an element obscured by another was failing with `wait for selector did not result in any nodes`. Now we properly detect this scenario and retry the action with different scroll positions. This affects all `Element` and `Frame` `click()`, `dblclick()`, `hover()`, `check()`, `uncheck()` and `tap()` actions.

Author: Does it make sense to add #264 also as a feature? We didn't have an auto-retry mechanism, and it's one of the most important features of PW.

Reviewer: :+1: Sure, go ahead.

Author: @imiric, how does it sound?

- Fixed `waitForSelector()` flakiness by retrying once on failure. ([#260](https://github.com/grafana/xk6-browser/pull/260))

- Fixed data race accessing `Frame.documentHandle` while switching execution contexts. ([#263](https://github.com/grafana/xk6-browser/pull/263))

- Fixed keyboard press of special keys. ([#283](https://github.com/grafana/xk6-browser/pull/283))

>   Previously pressing some keys like Backspace, Delete, ArrowLeft and AltGraph would panic because of a wrong CDP argument.

Author: I'm not sure whether users need to know about the wrong CDP argument?

```suggestion
  Previously pressing some keys like Backspace, Delete, ArrowLeft, and AltGraph would panic.
```

- Fixed negative metric emission in certain rare scenarios. ([#288](https://github.com/grafana/xk6-browser/pull/288))

- Fixed `Page.waitForFunction()`. ([#294](https://github.com/grafana/xk6-browser/pull/294))

>   This was previously broken and untested, and with the fix all polling variants (`requestAnimationFrame`, DOM mutation and interval) should work correctly. See [this example](https://github.com/grafana/xk6-browser/blob/v0.3.0/examples/waitforfunction.js) for details.

Author: Instead of using negative language (which may give a wrong impression), would it be better if we just say the following?

```suggestion
  `waitForFunction` wasn't reliable. This fix reimplements it from scratch and adds proper support for the polling options. The polling options can be one of `raf`, `mutation`, or a number (defaults to `raf`). See [this example](https://github.com/grafana/xk6-browser/blob/v0.3.0/examples/waitforfunction.js) for details.
```

Reviewer: It's not that `waitForFunction` wasn't reliable—it's that it didn't work at all (it would panic). It's also not a "from scratch" reimplementation, but more of a rewrite with a bunch of fixes.

But I'll change this to soften the tone and mention the proper valid options. :+1:

Reviewer: Let me know if ad0d4d9 resolves this.

Author: Nice write-up. Would it be better if we also add that polling defaults to `raf`?

Reviewer: Sure, done in 02ebb34.

- Fixed `Page.innerHTML()`, `Page.innerText()` and `Page.textContent()` panic. ([#299](https://github.com/grafana/xk6-browser/pull/299))

## Improvements

- Improved detection of non-clickable elements. ([#270](https://github.com/grafana/xk6-browser/pull/270))

  We now properly detect elements obscured by another element and return an error instead of waiting. This enables the fix in [#264](https://github.com/grafana/xk6-browser/pull/264).

- Cleanly close the WebSocket connection when `Browser.close()` is called. ([#268](https://github.com/grafana/xk6-browser/pull/268))

- Added lifecycle event validation to `Page.waitForLoadState()`. ([#300](https://github.com/grafana/xk6-browser/pull/300))

> - We have started updating our [JavaScript API documentation](https://k6.io/docs/javascript-api/xk6-browser/), which we know how important it is to serve as a reference while writing your scripts. This is an ardous task that won't be possible in a couple of weeks, but rest assured we're prioritizing it in all upcoming development cycles, and will be releasing gradual changes as they're done.

Reviewer: Should we mention and link to the [roadmap](https://github.com/grafana/xk6-browser/blob/241e40269e050624be21f94b80aca2cfed630d70/ROADMAP.md) here?

It's not really a code improvement, but neither are the docs... We don't link to it from anywhere, so people might miss it. Or should we link to it from the README instead?

Author: I was going to ask you the same :) 👍

Reviewer: So, what's your suggestion?

Author: 👍 = I agree, let's add it :)

Reviewer: As discussed over Slack, I added it both here (efcef63) and in the README (6e49bc8).
